### PR TITLE
libbpf-rs: optionally return full vecs from query

### PR DESCRIPTION
**Pull request by alexgartrell**
NOTE: this is totally API-breaking, but the initial API wasn't really useful.

This avoids allocation overhead and guessing by reading the values and re-invoking the system call. Each array is blocked by a feature flag from the iterator.

Here's an example program using iced-x86 to disassembly the jited code:

```
use iced_x86::Formatter;

fn main() {
    for prog in libbpf_rs::query::ProgInfoIter::default().include_jited_prog_insns(true) {
        let mut d = iced_x86::Decoder::new(32, &prog.jited_prog_insns, 0);
        let mut f = iced_x86::GasFormatter::new();
        while d.can_decode() {
            let insn = d.decode();
            let mut f_insn = String::new();
            f.format(&insn, &mut f_insn);
            println!("{}", f_insn);
        }
    }
}
```

### CODE REVIEW

**Comment by alexgartrell** [comment]
I don't think that this is a straightforward change so I'd consider this pull request more of an RFC. The tl;dr; is that it's really not valuable to just get the number of instructions if you actually want to introspect. The btf stuff is extra not-valuable for the same reason.

This probably ultimately requires actually writing out the code for the iterators instead of a big chunky macro to generate them, but I don't think that's the end of the world.

File: libbpf-rs/src/query.rs
Comment (anakryiko) [inline libbpf-rs/src/query.rs:271]: the actual loop looks unnecessary. You shouldn't need more than two calls for bpf_obj_get_info_by_fd: first to get lengths of necessary buffers, second to fill those buffers out by kernel. And then skip the second step if none of options were set. It should also simplify update_vec_field!() macro, probably, as you know that you are resizing vec.

File: libbpf-rs/src/query.rs
Comment (anakryiko) [inline libbpf-rs/src/query.rs:108]: a bit conflicted on removing len fields. They seem useful for cases when user doesn't use ProgInfoQueryOptions and wants to do something smarted and custom on their own after fetching lengths. But this might be a bit hypothethical problem... Having vec fields and separate lens doesn't seem like a big deal?

**Review by anakryiko — COMMENTED** [review]
I like these changes, as you said makes for much more usable API. I don't think breaking API change is a big deal, so I'd go ahead with this.

I left few comments on details (e.g., two explicit steps vs a loop), but those are minor. Let's wait for @danielocfb  to return to properly review this and land.

Thanks for improving libbpf-rs! :)

File: libbpf-rs/src/query.rs
Comment (alexgartrell) [reply libbpf-rs/src/query.rs:115]: In this model, the _len is just available by calling .len() on the resulting vector. So I do think it's worth it to just do it the clean way in this instance

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:116]: Why do we need this type and can't use https://docs.rs/libbpf-sys/1.2.1+v1.2.0/libbpf_sys/struct.bpf_line_info.html? Isn't that what it is?

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:226]: Have you tried expressing this as a function? It seems to me as if mutable references to objects could be passed in and that would suffice?

**Review by danielocfb — CHANGES_REQUESTED** [review]
Sorry for the delay. Left some comments. I'd really want for us to try and limit the number of macros; it is borderline unmaintainable. I don't think it's obvious what can be passed in to said macros and there is basically no contract that each argument has to fulfill.

File: libbpf-rs/src/query.rs
Comment (danielocfb) [reply libbpf-rs/src/query.rs:116]: Sorry, this comment was meant to apply to `LineInfoRaw`.

File: libbpf-rs/src/query.rs
Comment (danielocfb) [reply libbpf-rs/src/query.rs:115]: I have no strong feelings either way. Unless we have a good use case for the additional length fields this model of just the `Vec` carrying the length is fine with me. We label ourselves "opinionated eBPF tooling" after all :-)

**Comment by danielocfb** [comment]
> NOTE: this is totally API-breaking

I think that is fine.

File: libbpf-rs/src/query.rs
Comment (alexgartrell) [reply libbpf-rs/src/query.rs:115]: FWIW, upon reflection, the reason to do this would be to write an application that only takes the jited instructions if there are <= N of them (for example). This is likely niche enough to ignore but I'll let people argue with me in V2

File: libbpf-rs/src/query.rs
Comment (alexgartrell) [reply libbpf-rs/src/query.rs:128]: These are defined within macros in libbpf, so there's no clean way to pull it out.

https://www.kernel.org/doc/html/next/bpf/btf.html#:~:text=%23define%20BPF_LINE_INFO_LINE_NUM(line_col)%20%20%20%20%20%20%20%20((line_col)%20%3E%3E%2010)%0A%23define%20BPF_LINE_INFO_LINE_COL(line_col)%20%20%20%20%20%20%20%20((line_col)%20%26%200x3ff)

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:519]: Same comment as earlier about propagating errors up.

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:513]:
```suggestion
        let mut len = size_of_val(&item) as u32;
```

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:502]: Would you mind terribly documenting these changed members? We try to document everything that is new at the very least in the hopes that eventually we can remove all `#[allow(missing_docs)]`.

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:508]: Nit: let's import `std` items we use at the top for the sake of consistency with existing code:
```suggestion
        let mut item: libbpf_sys::bpf_btf_info = unsafe { mem::zeroed() };
```

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:534]: One more suggestion: how about we make `BtfInfo::name` a `CString`. This way we can forego this opinionated fallible conversion and let users choose how to handle errors and it's barely more than a `to_string_lossy` call on their end if they opt not to.
It's what we do in the existing BTF functionality: https://docs.rs/libbpf-rs/0.21.2/libbpf_rs/btf/struct.BtfType.html#method.name

**Review by danielocfb — COMMENTED** [review]
The updated changes look great! Thanks (and thanks for illustrating API usage in the example)! I left a few more minor comments.
Other than that I'd just hope we could add one or more unit tests for the changed functionality. Do you see a way to go about that? At least something like query a program and check non-zero instruction count or similar. Could also include parts of the example. It's mostly that we have something runnable in an automated fashion.

File: libbpf-rs/src/query.rs
Comment (alexgartrell) [reply libbpf-rs/src/query.rs:534]: For what it's worth, ProgramInfo::name is a String as well. Should that be changed?

Comment (danielocfb) [reply libbpf-rs/src/query.rs:534]: Ah, I missed that. Yes, that would be great. Though I won't insist you do it as you didn't really touch that member. So I'll leave it up to you. I added it to my list of things to do as well.

File: examples/bpf_query/Cargo.toml
Comment (alexgartrell) [reply examples/bpf_query/Cargo.toml:14]: It's a runtime dependency for the example in particular. Not a new dependency for libbpf-rs

File: libbpf-rs/src/query.rs
Comment (alexgartrell) [reply libbpf-rs/src/query.rs:513]: Fixing this instance, but there are others in this file for a subsequent diff

File: examples/bpf_query/Cargo.toml
Comment (danielocfb) [reply examples/bpf_query/Cargo.toml:14]: Ah yes, I somehow read this as `libbpf-rs`'s `Cargo.toml`.

**Comment by alexgartrell** [comment]
I agree the lack of testing is a problem, but I propose we provide it in another diff. The real path here is to write cfg(test) stubs for the libbpf calls and mock it up completely, but that will be a monster change.

Absent that, you end up with tests that need sudo, which isn't super necessary/desirable here IMHO

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:573]: Let's please not print stuff in a library. It will end up in callers without any context in places where it is unlikely to be expected.

File: libbpf-rs/src/query.rs
Comment (danielocfb) [inline libbpf-rs/src/query.rs:503]: Hm, I thought we agreed on making this `CString`. But now not even the comment shows up anymore...

**Review by danielocfb — APPROVED** [review]
Seems mostly fine now. I don't think you addressed all the points or perhaps I dreamed about pointing them out or GitHub swallowed them. Will create a fixup commit and add a CHANGELOG entry, but merging this change. Thanks!

**Comment by danielocfb** [comment]
> Absent that, you end up with tests that need sudo, which isn't super necessary/desirable here IMHO

Half our tests already require `sudo`.

**Pull request merged**
